File: backend/src/main/java/de/damirutje/rockpaperscissors/service/GameService.java
package de.damirutje.rockpaperscissors.service;

import de.damirutje.rockpaperscissors.dto.GameStartDto;
import de.damirutje.rockpaperscissors.model.Game;
import de.damirutje.rockpaperscissors.model.HandSign;

public interface GameService {

    /**
     * Creates and starts a new {@link Game}.
     * @return id of created {@link Game}
     */
    long startGame();

    /**
     * Creates and starts a new {@link Game} by specified settings.
     * @param gameStartDto settings for new {@link Game}.
     * @return id of created {@link Game}
     */
    long startGame(GameStartDto gameStartDto);

    /**
     * Requests {@link Game} entity from database by specified id.
     * @param id of requested {@link Game}
     * @return requested {@link Game}
     */
    Game getGame(long id);

    /**
     * Makes and persists the move for requested {@link Game} entity.
     * @param gameId id of requested {@link Game}
     * @param handSign to make move with
     * @return {@link Game} after move
     */
    Game makeMove(long gameId, HandSign handSign);

    /**
     * Sets game state for requested {@link Game} entity to 'Aborted'.
     * @param id of requested {@link Game} to abort
     */
    void abortGame(long id);
}

File: frontend/src/app/shared/game.service.ts
import { HttpClient, HttpHeaders } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable, of } from "rxjs";
import { catchError } from 'rxjs/operators';
import { environment } from "../../environments/environment";
import { GameStartDto } from "./game-start-dto.model";
import { Game } from "./game.model";

@Injectable()
export class GameService {

  private gameUrl = `${environment.baseUrl}/game`;  // URL to web api

  httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
  };

  constructor(private http: HttpClient) { }

  /** Start a new game with given game settings */
  startGame(game?: GameStartDto): Observable<any> {
    const httpOptions = {
      headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
      observe: "response" as 'body',
    };
    return this.http.post(`${this.gameUrl}/start`, game, httpOptions);
  }

  /** Get game by id */
  getGame(id: number): Observable<Game> {
    return this.http.get<Game>(`${this.gameUrl}/${id}`);
  }

  /** Make move for given game */
  makeMove(id: number, handSign: string): Observable<Game> {
    return this.http.post<Game>(`${this.gameUrl}/${id}/move`, JSON.stringify(handSign), this.httpOptions);
  }

  /** Abort game */
  abortGame(id: number): Observable<any> {
    return this.http.post(`${this.gameUrl}/${id}/abort`, this.httpOptions);
  }
}
